src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3957,112 +3951,196 @@ private void CreateInstanceCheckThis()
+                        Debug.Assert(!rt.IsGenericCOMObjectImpl(), "__ComObject base class should've been blocked.");
+                        defaultCtorRMH = default; // ignore any parameterless ctor
+                    }
+                    else
+                    {
+                        if (!_pMT->HasDefaultConstructor)
+                        {
+                            throw new MissingMethodException(SR.Format(SR.Arg_NoDefCTor, rt));
+                        }
+
+                        // Reference type with explicit parameterless ctor typed as (object) -> void.
+                        // We'll pass the actual ctor address as the state object.
+
+                        static void ReferenceTypeStub(object? @this, IntPtr state)
+                        {
+                            ((delegate*<object, void>)state)(@this!);
+                        }
+                        _pfnCtorStub = &ReferenceTypeStub;
+                        _ctorStubState = RuntimeMethodHandle.GetFunctionPointer(defaultCtorRMH);
+                    }
+                }
+
+                if (defaultCtorRMH.IsNullHandle())
+                {
+                    CtorIsPublic = true; // implicit parameterless ctor is always considered public
+                }
+                else
+                {
+                    CtorIsPublic = (RuntimeMethodHandle.GetAttributes(defaultCtorRMH) & MethodAttributes.Public) != 0;
+                }
+
+                Debug.Assert(_pfnNewobj != null);
+                Debug.Assert(_pfnCtorStub != null); // we use null singleton pattern if no ctor call is necessary
             }
 
-            if (canBeCached && fillCache)
+            internal bool CtorIsPublic { get; }
+
+            private static void CtorNoopStub(object? @this, IntPtr state) { }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal object? CreateUninitializedObject(RuntimeType rt)
             {
-                // cache the ctor
-                GenericCache = new ActivatorCache(runtimeCtor);
+                // We don't use RuntimeType, but we force the caller to pass it so
+                // that we can keep it alive on their behalf. Once the object is
+                // constructed, we no longer need the reference to the type instance,
+                // as the object itself will keep the type alive.
+
+#if DEBUG
+                Debug.Assert(rt == _originalRT, "Caller passed the wrong RuntimeType to this routine.");
+#endif
+
+                object? retVal = _pfnNewobj(_pMT);
+                GC.KeepAlive(rt);
+                return retVal;
             }
 
-            return instance;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal void CallCtorOverUninitializedObject(object? newObj) => _pfnCtorStub(newObj, _ctorStubState);
+        }
+
+        // This method mimics the overload present in the mono codebase. It allows shared source callers
```

Comment:
No point in keeping this around. We should update or remove shared source callers as necessary.

Reply:
(Or move it to Mono-specific file if it is hard to do.)

Reply:
In the coreclr code base I moved these checks (block "System.Void", etc.) into the unified native function, so they always run now. I think it'll be possible to do the same on the mono side, that way I can remove this stub and have the shared callers call the simpler overloads.